File: android/src/main/kotlin/io/devcrew/snapchatloginkit/model/UserDataQuery.kt
package io.devcrew.snapchatloginkit.model

import com.snap.loginkit.BitmojiQuery;
import com.snap.loginkit.UserDataQuery;

data class UserFetchDataQuery(
    val isWithDisplayName: Boolean,
    val isWithExternalId: Boolean,
    val isWithIdToken: Boolean,
    val isWithProfileLink: Boolean,
    val isWithBitmojiAvatarId: Boolean,
    val isWithBitmojiAvatarUrl: Boolean
) {
    constructor(map: Map<String, Any?>) : this(
        isWithDisplayName = map["isWithDisplayName"] as? Boolean ?: false,
        isWithExternalId = map["isWithExternalId"] as? Boolean ?: false,
        isWithIdToken = map["isWithIdToken"] as? Boolean ?: false,
        isWithProfileLink = map["isWithProfileLink"] as? Boolean ?: false,
        isWithBitmojiAvatarId = map["isWithBitmojiAvatarId"] as? Boolean ?: false,
        isWithBitmojiAvatarUrl = map["isWithBitmojiAvatarUrl"] as? Boolean ?: false
    )

    fun prepareUserDataQuery(): UserDataQuery {
        val builder = UserDataQuery.newBuilder()
        val bitmojiQueryBuilder: BitmojiQuery.Builder = BitmojiQuery.newBuilder()
        val bitmojiQuery: BitmojiQuery
        if (isWithDisplayName) {
            builder.withDisplayName()
        }
        if (isWithExternalId) {
            builder.withExternalId()
        }
        if (isWithIdToken) {
            builder.withIdToken()
        }

        if (isWithBitmojiAvatarId || isWithBitmojiAvatarUrl) {
            if (isWithBitmojiAvatarId) {
                bitmojiQueryBuilder.withAvatarId()
            }
            if (isWithBitmojiAvatarUrl) {
                bitmojiQueryBuilder.withTwoDAvatarUrl()
            }
            bitmojiQuery = bitmojiQueryBuilder.build()
            builder.withBitmoji(bitmojiQuery)
        }
        return builder.build()
    }
}

File: ios/Classes/Resources/UserDataQuery.swift
//
//  UserDataQuery.swift
//  snapchat_loginkit
//
//  Created by Najam Us Saqib on 23/02/2024.
//

import Foundation
import SCSDKLoginKit

// MARK: - UserFetchDataQuery -

class UserFetchDataQuery {
    
    // MARK: - Properties -
    
    let isWithDisplayName: Bool
    let isWithExternalId: Bool
    let isWithIdToken: Bool
    let isWithProfileLink: Bool
    let isWithBitmojiAvatarId: Bool
    let isWithBitmojiAvatarUrl: Bool
    
    init(fromMap map: [String : Any]) {
        isWithDisplayName = map["isWithDisplayName"] as? Bool ?? false
        isWithExternalId = map["isWithExternalId"] as? Bool ?? false
        isWithIdToken = map["isWithIdToken"] as? Bool ?? false
        isWithProfileLink = map["isWithProfileLink"] as? Bool ?? false
        isWithBitmojiAvatarId = map["isWithBitmojiAvatarId"] as? Bool ?? false
        isWithBitmojiAvatarUrl = map["isWithBitmojiAvatarUrl"] as? Bool ?? false
    }
    
    
    // MARK: - Functions -
    
    func prepareUserDataQuery() -> SCSDKUserDataQuery {
        let builder = SCSDKUserDataQueryBuilder()
        
        if (isWithDisplayName) {
            builder.withDisplayName()
        }
        if (isWithExternalId) {
            builder.withExternalId()
        }
        if (isWithIdToken) {
            builder.withIdToken()
        }
        if (isWithProfileLink) {
            builder.withProfileLink()
        }
        if (isWithBitmojiAvatarId) {
            builder.withBitmojiAvatarID()
        }
        if (isWithBitmojiAvatarUrl) {
            builder.withBitmojiTwoDAvatarUrl()
        }
        
        return builder.build()
    }
}
